### PR TITLE
Abandon pip in python3-pep517 build style

### PR DESCRIPTION
**Pull request by ahesford**
Using `pip`, the package manager, as a builder and installer in the `python3-pep517` build style was always a poor choice. Now that `python3-build` and `python3-installer` have apparently matured a bit and fall under the PyPA umbrella, let's use those instead.

This poses a bit of a bootstrap problem because PyPA doesn't seem to worry about dependency cycles, so we need to handle the core packages apart from the `python3-pep517` build style. Thus, we still use `pip` to install `python3-installer` and `python3-flit_core`.

Note that these packages are still in flux. For example, `pyproject-hooks` used to be called `pep517` and is still named that in the released version, but I'm using the new name to avoid a package rename in the future. Also, `python3-build` appears to have moved to `python3-flit_core` as the build backend, but the released version is still using `python3-setuptools`. We'll get there eventually...

I'm going to rebuild all of the `python3-pep517` packages to make sure this doesn't break anything.

### CODE REVIEW

**Comment by eli-schwartz** [comment]
I still think it's viable to have a flit build style.

Question: is it viable to have `installer` download both an sdist and a wheel, then install the wheel without needing to depend on flit? This would allow flit_core to depend on `installer` instead of pip. In fact it could use its own build style as it will be on its own import path, you'd just need to skip adding an extra copy of flit to the makedepends.

**Comment by ahesford** [comment]
The flit build style doesn't help us here because we would still need something to install wheels. That means that, except for `python3-build` and `python3-pyproject-hooks`, every other package added here will still need custom behavior. Any package *not* required to bootstrap the PEP-517 toolchain can just use flit with the normal style.

We tend not to fetch prebuilt artifacts, so grabbing the wheel isn't really viable. If pip becomes completely unserviceable, I guess I could pull the `installer` sdist into the `python3-flit_core` template and use a local, uninstalled version to do the installation there. For now, pip will get the job done.

**Comment by eli-schwartz** [comment]
> The flit build style doesn't help us here because we would still need something to install wheels. That means that, except for `python3-build` and `python3-pyproject-hooks`, every other package added here will still need custom behavior.

Yeah, all two of them. :p To put it another way, you reduce the problem by 50%.

> We tend not to fetch prebuilt artifacts, so grabbing the wheel isn't really viable.

The official PyPA policy on that is "wheels are just source code copied into a different directory", which is... sort of kind of true, if you squint.

Simply hope and pray that you don't ever need to apply patches to it, as it then rapidly becomes a built artifact (where the patches are part of the input pipeline) as you cannot patch a wheel.

(Yes I know this isn't nearly as helpful as PyPA thinks. We've finally reached a nearly bootstrappable world where you only need to treat a single package as magic, or, depending, two packages. But they really want you to treat *something* as magic.)

**Comment by ahesford** [comment]
This is also a social engineering problem; having a build style means that somebody will use it. We shouldn't fragment the build styles further with two build styles that do essentially the same thing for packages that use `flit_core`, so it's preferable to just repeat the logic in these few bootstrap packages.

**Comment by ahesford** [comment]
@eli-schwartz you at least convinced me to drop pip from `python3-flit_core`.

**Comment by eli-schwartz** [comment]
Well, I'll take my victories where I can get them. :D Less pip in build recipes is always a good thing.

Maybe when I complete the work on a PEP 517 build backend inside Meson, I can just contribute meson.build files for all PyPA core infrastructure, and then you can switch to the meson build style... since Meson includes both a builder and an installer. :D 

(The eventual goal is that Meson itself is built via a custom recipe to do `./meson.py setup builddir && DESTDIR=${DESTDIR} ./meson.py install -C builddir/` and still provides all the same metadata that setuptools or flit or `$build_tool_of_the_day` provides. Zero dependencies.)

**Comment by ahesford** [comment]
Having meson take care of building and installation without also trying to automatically manage dependencies like pip would be a great option.

**Comment by ahesford** [comment]
With the extra commits added to this branch, every pep517 package now builds properly.

File: srcpkgs/python3-subprocess-tee/template [hunk 1]
```diff
@@ -15,7 +15,7 @@ checksum=ff5cced589a4b8ac973276ca1ba21bb6e3de600cde11a69947ff51f696efd577
 make_check=no
 
 post_patch() {
-	vsed -e "/setuptools_scm_git_archive/d" -i pyproject.toml
+	vsed -e "/setuptools_scm_git_archive/d" -e "/pip/d" -i pyproject.toml
```
Comment (eli-schwartz) [inline srcpkgs/python3-subprocess-tee/template:18]: Holy what the cow. Who on earth specifies that pip needs to be installed inside the virtualenv that pip creates for building?

File: srcpkgs/python3-pendulum/template [hunk 2]
```diff
@@ -14,6 +15,12 @@ changelog="https://github.com/sdispater/pendulum/blob/master/CHANGELOG.md"
 distfiles="https://github.com/sdispater/pendulum/archive/refs/tags/${version}.tar.gz"
 checksum=4185efa48358cb9dca57a6afbc471d4b260e70f56609d5030197373564ac4a07
 
+do_build() {
+	mkdir -p build
+	TMPDIR="${PWD}/build" python3 -m pip wheel --no-deps --no-clean \
+		--use-pep517 --no-build-isolation --wheel-dir dist .
+}
```
Comment (eli-schwartz) [inline srcpkgs/python3-pendulum/template:22]: This one is great. `python -m build` attempts to run poetry's undocumented `build.py` plugin.

I've always been pretty meh about the use of `python -m` anything unless the goal is very specifically to inject $PWD into the python module search path, which granted you may sometimes want to do (pytest, specifically).

`/usr/bin/pyproject-build` arguably works just as well.

Comment (ahesford) [reply srcpkgs/python3-pendulum/template]: Thanks for catching this; it completely escaped me!

**Comment by eli-schwartz** [comment]
https://github.com/void-linux/void-packages/blob/208681d556e650549447f32276e1b645b44c5b5b/srcpkgs/python3-Sphinx/template#L6

Missed one needed change to add dist/, /cc @q66

**Pull request merged**